geps/gep-1731/index.md

```diff
@@ -0,0 +1,409 @@
+
+    // Backoff specifies the minimum duration a Gateway should wait between
+    // retry attempts and is represented in Gateway API Duration formatting.
+    //
+    // For example, setting the `rules[].retry.backoff` field to the value
+    // `100ms` will cause a backend request to first be retried approximately
+    // 100 milliseconds after timing out or reciveing a response code configured
+    // to be retryable.
+    //
+    // An implementation MAY use an exponential or alternative backoff strategy
+    // for subsequent retry attempts, MAY cap the maximum backoff duration to
+    // some amount greater than the specified minimum, and MAY add arbitrary
+    // jitter to stagger requests, as long as unsuccessful backend requests are
+    // not retried before the configured minimum duration.
+    //
+    // If a Request timeout (`rules[].timeouts.request`) is configured on the
+    // route, the entire duration of the initial request and any retry attempts
+    // MUST not exceed the Request timeout duration. If any retry attempts are
+    // still in progress when the Request timeout duration has been reached,
+    // these SHOULD be cancelled if possible and the Gateway MUST immediately
+    // return a timeout error.
+    //
+    // If a BackendRequest timeout (`rules[].timeouts.backendRequest`) is
+    // configured on the route, any retry attempts which reach the configured
+    // BackendRequest timeout duration without a response SHOULD be cancelled if
+    // possible and the Gateway should wait for at least the specified backoff
+    // duration before attempting to retry the backend request again.
+    //
+    // If a BackendRequest timeout is _not_ configured on the route, retry
+    // attempts MAY time out after an implementation default duration, or MAY
+    // remain pending until a configured Request timeout or implementation
+    // default duration for total request time is reached.
+    //
+    // When this field is unspecified, the time to wait between retry attempts
+    // is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Backoff *Duration `json:"backoff,omitempty"`
+}
+
+// HTTPRouteRetryStatusCode defines an HTTP response status code for
+// which a backend request should be retried.
+//
+// Implementations MUST support the following status codes as retryable:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// Implementations SHOULD support the `"5xx"` shorthand for matching any
+// status code in the 500-599 range, and SHOULD support specifying any
+// status code within the valid 100-599 (inclusive) range.
+//
+// Implementations MAY support values in the 600-999 (inclusive) range,
+// which are not valid for HTTP clients, but are sometimes used for
+// communicating application-specific errors, or additional shorthand 
+// codes for any `[0-9]xx` range.
+//
+// +kubebuilder:validation:Pattern=`^[1-9](?:[0-9][0-9]|xx)$`
+type HTTPRouteRetryStatusCode string
```

Comment:
As we've seen in upstream Kubernetes with the `intOrString` type, this can unfortunately get rather messy. I think it may be safer to have a struct where you can specify either a string `name` or int `code`.

Reply:
This is a feature that **A**na (the **a**pplication dev) is likely to be dealing with. From Ana's point of view, this is probably ideal:

```yaml
retry:
  codes: [ 500, 502, connect-failure ]
  attempts: 5
  backoff: 100ms
```

That will feel natural to Ana, but it requires `intOrString` and breaks structural schemas. 🙁 Having a simple string here, as @mikemorris does, is probably the next simplest thing, again from Ana's PoV:

```yaml
retry:
  codes: [ "500", "502", "connect-failure" ]
  attempts: 5
  backoff: 100ms
```

She may be annoyed at having to remember to quote things, but it's consistent and it gets validated at apply time... and it's _very_ hard to come up with a way that it benefits _Ana_ to make her do

```yaml
retry:
  codes:
    - status: 500
    - status: 502
    - name: connect-failure
  attempts: 5
  backoff: 100ms
```

Reply:
I really dislike mixed string + int types, but especially since we're likely going to have "5xx" as not just a supported value, but potentially the _only_ portable value, I think I could get behind strings here.

Reply:
Resolving this planning to stick to all string values as currently proposed.